File: Functions/computeCheckupCurves.m
function checkups = computeCheckupCurves(segments, selectedTime, selectedVoltage, selectedCurrent, selectedTimeS, windowSize)
% computeCheckupCurves - Compute per-checkup OCP / capacity / dQ-dV curves.
%
% This is the single, authoritative implementation of the checkup-discharge
% math that was previously duplicated between:
%   (a) Functions/analyzeCheckupDischarge.m  (used by ExtractAgeingData.m), and
%   (b) the inline "Recompute per-checkup curves" block in PlotCellSummary.m.
%
% It performs NO plotting and NO file I/O; it only returns the numeric curves
% so that each caller can render / export them however it needs. Callers that
% also need State-of-Charge / OCV interpolation, Full-Equivalent-Cycles, or
% figures build those on top of the returned data.
%
% Acceptance test (identical to the previous implementations): a segment is a
% valid checkup discharge only if it both STARTS above 4.1 V and ENDS below
% 2.76 V (i.e. a complete top-to-bottom constant-current discharge).
%
% Inputs:
%   segments        - Cell array of segment index ranges (into the selected*
%                     arrays), e.g. from findCheckupSegments
%   selectedTime    - datetime array for the selected range
%   selectedVoltage - voltage array for the selected range [V]
%   selectedCurrent - current array for the selected range [A]
%   selectedTimeS   - seconds-since-start array for the selected range [s]
%   windowSize      - movmean window for the dQ/dV smoothing pipeline
%
% Output:
%   checkups - struct array (1 x nValid), one element per valid checkup, with:
%       .segmentNumber   - index of this segment within the input `segments`
%                          cell array (preserved so callers can reproduce the
%                          original age-ordered colour: [i/N, 0, 1-i/N])
%       .segmentIndices  - the segment's index vector into the selected* arrays
%       .capacity_Ah_vec - per-sample discharge capacity vector [Ah]
%                          (= -cumtrapz(t, I)/3600)
%       .voltage_vec     - per-sample voltage vector [V]
%       .dQdV_vec        - smoothed dQ/dV vector [As/V]
%       .startTime       - datetime of the first sample of the segment
%       .endTime         - datetime of the last sample of the segment
%       .capacity_Ah     - scalar discharge capacity [Ah] (= -min(cumtrapz/3600))
%       .timeStamp       - datetime used for trending (segment start time)
%   If no valid checkup is found, `checkups` is a 0x0 struct array carrying the
%   same field names (so numel(checkups)==0 and field access is safe).
%
% Author: GitHub Copilot (for Róbinson Medina / Feye Hoekstra)
% Date:   2026-07-28
%
% Compliance: R-003 (R2022a-safe), R-006 (Ah/V/s units), R-012 (commented).

% Preallocate an empty struct array carrying every field, so callers can
% always rely on the field set and on numel() even when nothing is found.
checkups = struct('segmentNumber', {}, 'segmentIndices', {}, ...
    'capacity_Ah_vec', {}, 'voltage_vec', {}, 'dQdV_vec', {}, ...
    'startTime', {}, 'endTime', {}, 'capacity_Ah', {}, 'timeStamp', {});

if isempty(segments)
    return
end

for i = 1:numel(segments)
    segmentIndices = segments{i};
    sV  = selectedVoltage(segmentIndices);          % segment voltage [V]
    sI  = selectedCurrent(segmentIndices);          % segment current [A]
    sT  = selectedTime(segmentIndices);             % segment datetime
    sTs = selectedTimeS(segmentIndices);            % segment time [s]
    sTs = sTs - sTs(1);                             % normalise to start at 0 s

    % Acceptance test: only complete discharge cycles (start > 4.1 V, end < 2.76 V).
    if ~(sV(end) < 2.76 && sV(1) > 4.1)
        continue
    end

    % Per-sample discharge capacity [Ah]. Discharge current is negative, so the
    % negated integral is a positive, increasing capacity trace.
    capVec = -cumtrapz(sTs, sI) / 3600;

    % Smoothed dQ/dV via the movmean/gradient pipeline (identical smoothing to
    % the previous implementations): smooth I and V, integrate charge, then take
    % gradient(charge)/gradient(V) and smooth again.
    smI  = movmean(sI, windowSize);
    smV  = movmean(sV, windowSize);
    dQdV = movmean(gradient(cumtrapz(sTs, smI)) ./ gradient(smV), windowSize);

    % Assemble this checkup's record.
    entry = struct();
    entry.segmentNumber   = i;                       % original position in `segments`
    entry.segmentIndices  = segmentIndices;
    entry.capacity_Ah_vec = capVec;
    entry.voltage_vec     = sV;
    entry.dQdV_vec        = dQdV;
    entry.startTime       = sT(1);
    entry.endTime         = sT(end);
    entry.capacity_Ah     = -min(cumtrapz(sTs, sI) / 3600);  % scalar checkup capacity
    entry.timeStamp       = sT(1);

    checkups(end+1) = entry; %#ok<AGROW> % number of checkups is small (tens)
end

end

File: Functions/computeCheckupCurves.py
"""
computeCheckupCurves - Compute per-checkup OCP / capacity / dQ-dV curves.

Python counterpart of Functions/computeCheckupCurves.m (todo #021). Single,
authoritative implementation of the checkup-discharge math shared by
Functions/analyzeCheckupDischarge.py and any future PlotCellSummary.py port.
Performs NO plotting and NO file I/O; callers add FEC/SoC/OCV interpolation
and plotting on top.

Acceptance test (identical to the MATLAB implementation): a segment is a
valid checkup discharge only if it STARTS above 4.1 V and ENDS below 2.76 V.

Author: GitHub Copilot (for Robinson Medina / Feye Hoekstra)
Date: 2026-08-24

Compliance: R-004 (Python 3.9+), R-006 (Ah/V/s units), R-012 (commented).
"""

import numpy as np
import pandas as pd

try:
    # scipy >= 1.6 name (cumtrapz is deprecated/removed in newer releases).
    from scipy.integrate import cumulative_trapezoid as _cumtrapz
except ImportError:  # pragma: no cover - older scipy fallback
    from scipy.integrate import cumtrapz as _cumtrapz


def _moving_average(data, window_size):
    """Centered moving average matching MATLAB's movmean (edge-shrinking)."""
    return pd.Series(data).rolling(window=window_size, center=True, min_periods=1).mean().values


def compute_checkup_curves(segments, selected_time, selected_voltage, selected_current,
                            selected_time_s, window_size):
    """
    Compute per-checkup OCP / capacity / dQ-dV curves for every valid segment.

    Parameters
    ----------
    segments : list of numpy.ndarray
        Index ranges (into the selected_* arrays) of candidate checkup segments,
        e.g. from find_checkup_segments.
    selected_time : numpy.ndarray of datetime64
        Datetime array for the selected range.
    selected_voltage : numpy.ndarray
        Voltage array for the selected range [V].
    selected_current : numpy.ndarray
        Current array for the selected range [A].
    selected_time_s : numpy.ndarray
        Seconds-since-start array for the selected range [s].
    window_size : int
        movmean window for the dQ/dV smoothing pipeline.

    Returns
    -------
    checkups : list of dict
        One dict per valid checkup (empty list if none found), each with keys:
        'segment_number', 'segment_indices', 'capacity_Ah_vec', 'voltage_vec',
        'dQdV_vec', 'start_time', 'end_time', 'capacity_Ah', 'time_stamp'
        (mirrors the MATLAB struct-array field names).
    """

    checkups = []
    if not segments:
        return checkups

    for i, segment_indices in enumerate(segments):
        s_v = selected_voltage[segment_indices]          # segment voltage [V]
        s_i = selected_current[segment_indices]           # segment current [A]
        s_t = selected_time[segment_indices]               # segment datetime
        s_ts = selected_time_s[segment_indices]            # segment time [s]
        s_ts = s_ts - s_ts[0]                              # normalise to start at 0 s

        # Acceptance test: only complete discharge cycles (start > 4.1 V, end < 2.76 V).
        if not (s_v[-1] < 2.76 and s_v[0] > 4.1):
            continue

        # Per-sample discharge capacity [Ah]; discharge current is negative,
        # so negating the integral gives a positive, increasing capacity trace.
        cap_vec = -_cumtrapz(s_i, s_ts, initial=0) / 3600

        # Smoothed dQ/dV via the movmean/gradient pipeline (identical
        # smoothing to the MATLAB implementation): smooth I and V, integrate
        # charge, then take gradient(charge)/gradient(V) and smooth again.
        sm_i = _moving_average(s_i, window_size)
        sm_v = _moving_average(s_v, window_size)
        charge = _cumtrapz(sm_i, s_ts, initial=0)
        dv = np.gradient(sm_v)
        dv[dv == 0] = np.nan  # avoid division by zero, matches MATLAB Inf->NaN-safe behaviour
        dqdv = _moving_average(np.gradient(charge) / dv, window_size)

        checkups.append({
            'segment_number': i,                       # 0-based position in `segments`
            'segment_indices': segment_indices,
            'capacity_Ah_vec': cap_vec,
            'voltage_vec': s_v,
            'dQdV_vec': dqdv,
            'start_time': s_t[0],
            'end_time': s_t[-1],
            'capacity_Ah': -np.min(_cumtrapz(s_i, s_ts, initial=0) / 3600),  # scalar capacity
            'time_stamp': s_t[0],
        })

    return checkups


__all__ = ['compute_checkup_curves']
